python/lsst/ap/association/filterDiaSourceCatalog.py

```diff
@@ -0,0 +1,117 @@
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+__all__ = (
+    "FilterDiaSourceCatalogConfig",
+    "FilterDiaSourceCatalogTask",
+)
+
+import lsst.pex.config as pexConfig
+import lsst.pipe.base as pipeBase
+import lsst.pipe.base.connectionTypes as connTypes
+from lsst.utils.timer import timeMethod
+from lsst.afw.table import SourceCatalog
+import numpy as np
+
+
+class FilterDiaSourceCatalogConnections(
+    pipeBase.PipelineTaskConnections,
+    dimensions=("instrument", "visit", "detector"),
+    defaultTemplates={"coaddName": "deep", "fakesType": ""},
+):
+    """Connections class for FilterDiaSourceCatalogTask."""
+
+    diaSourceCat = connTypes.Input(
+        doc="Catalog of DiaSources produced during image differencing.",
+        name="{fakesType}{coaddName}Diff_diaSrc",
+        storageClass="SourceCatalog",
+        dimensions=("instrument", "visit", "detector"),
+    )
+
+    filteredDiaSourceCat = connTypes.Output(
+        doc="Output catalog of DiaSources after filtering.",
+        name="{fakesType}{coaddName}Diff_candidateDiaSrc",
+        storageClass="SourceCatalog",
+        dimensions=("instrument", "visit", "detector"),
+    )
+
+    rejectedDiaSources = connTypes.Output(
+        doc="Optional output storing all the rejected DiaSources.",
+        name="{fakesType}{coaddName}Diff_rejectedDiaSrc",
+        storageClass="SourceCatalog",
+        dimensions={"instrument", "visit", "detector"},
+    )
+
+    def __init__(self, *, config=None):
+        super().__init__(config=config)
+        if not self.config.doWriteSkySources:
+            self.outputs.remove("rejectedDiaSources")
+
+
+class FilterDiaSourceCatalogConfig(
+    pipeBase.PipelineTaskConfig, pipelineConnections=FilterDiaSourceCatalogConnections
+):
+    """Config class for FilterDiaSourceCatalogTask."""
+
+    doRemoveSkySources = pexConfig.Field(
+        dtype=bool,
+        default=False,
+        doc="Input DiaSource catalog contains SkySources that should be "
+        "removed before storing the output DiaSource catalog.",
+    )
+
+    doWriteSkySources = pexConfig.Field(
+        dtype=bool,
+        default=True,
+        doc="Store the output DiaSource catalog containing all the rejected "
+        "sky sources."
+    )
+
+
+class FilterDiaSourceCatalogTask(pipeBase.PipelineTask):
+    """Filter out sky sources from a DiaSource catalog."""
+
+    ConfigClass = FilterDiaSourceCatalogConfig
+    _DefaultName = "filterDiaSourceCatalog"
+
+    @timeMethod
+    def run(self, diaSourceCat):
+        """Filter sky sources from the supplied DiaSource catalog.
+
+        Parameters
+        ----------
+        diaSourceCat : `lsst.afw.table.SourceCatalog`
+            Catalog of sources measured on the difference image.
+
+        Returns
+        -------
+        results : `lsst.pipe.base.Struct`
+            Results struct with components.
+        """
+        rejectedSkySources = SourceCatalog()
```

Comment:
I don't think this will work if `doRemoveSkySources` is False, since it won't have the right schema and would probably raise an error if you tried to write it to the Butler. You overwrite it later if we do remove sky sources, so just set it to `None` here and check if it is `None` before the return. If it is, set it to `rejectedSkySources = SourceCatalog(diaSourceCat.getSchema())`.